packages/components/src/components/SquiggleChart.tsx

```diff
@@ -298,28 +55,28 @@ export const SquiggleChart: React.FC<SquiggleChartProps> = ({
   showControls = false,
   chartSettings = defaultChartSettings,
 }) => {
-  let expressionResult = run(squiggleString, bindings, environment, jsImports);
-  if (expressionResult.tag !== "Ok") {
-    return (
-      <ErrorAlert heading={"Parse Error"}>
-        {errorValueToString(expressionResult.value)}
-      </ErrorAlert>
-    );
+  const { result } = useSquiggle({
+    code: squiggleString,
+    bindings,
+    environment,
+    jsImports,
+    onChange,
+  });
+
+  if (result.tag !== "Ok") {
+    return <SquiggleErrorAlert error={result.value} />;
   }
 
-  let e = environment ?? defaultEnvironment;
-  let expression = expressionResult.value;
-  onChange(expression);
   return (
     <SquiggleItem
-      expression={expression}
+      expression={result.value}
       width={width}
       height={height}
       showSummary={showSummary}
       showTypes={showTypes}
       showControls={showControls}
       chartSettings={chartSettings}
-      environment={e}
+      environment={environment ?? defaultEnvironment}
```

Comment:
Today I learnt something! This is certainly something I'd love to see being used more often.